New GeoHexGrid aggregation (#82924)
elastic/elasticsearch: docs/reference/aggregations/bucket/geohexgrid-aggregation.asciidoc

[role="xpack"]
[[search-aggregations-bucket-geohexgrid-aggregation]]
=== Geohex grid aggregation
++++
<titleabbrev>Geohex grid</titleabbrev>
++++

A multi-bucket aggregation that groups <<geo-point,`geo_point`>>
values into buckets that represent a grid.
The resulting grid can be sparse and only
contains cells that have matching data. Each cell corresponds to a
https://h3geo.org/docs/core-library/h3Indexing#h3-cell-indexp[H3 cell index] a
and labeled using the https://h3geo.org/docs/core-library/h3Indexing#h3index-representation
[H3Index representation].

* High precision keys have a larger range for x and y, and represent tiles that
cover only a small area.
* Low precision keys have a smaller range for x and y, and represent tiles that

> each cover a large area.

Author: copy and paste bug ... I removed all together as you cannot really infer information easily from the key.

{"index":{"_id":3}}
{"location": "52.371667,4.914722", "name": "Nederlands Scheepvaartmuseum"}
{"index":{"_id":4}}
{"location": "51.222900,4.405200", "name": "Letterenhuis"}
{"index":{"_id":5}}
{"location": "48.861111,2.336389", "name": "Musée du Louvre"}
{"index":{"_id":6}}
{"location": "48.860000,2.327000", "name": "Musée d'Orsay"}

POST /museums/_search?size=0
{
  "aggregations": {
    "large-grid": {
      "geohex_grid": {
        "field": "location",
        "precision": 4
      }
    }
  }
}
--------------------------------------------------

Response:

[source,console-result]
--------------------------------------------------
{
  ...
  "aggregations": {
    "large-grid": {
      "buckets": [
        {
          "key": "841969dffffffff",
          "doc_count": 3
        },
        {
          "key": "841fb47ffffffff",
          "doc_count": 2
        },
        {
          "key": "841fa4dffffffff",
          "doc_count": 1
        }
      ]
    }
  }
}
--------------------------------------------------
// TESTRESPONSE[s/\.\.\./"took": $body.took,"_shards": $body._shards,"hits":$body.hits,"timed_out":false,/]

> 

Reviewer: I'd insert the parameter definitions (currently the `Options` section) here. I'd also add a section immediately after with the response body definitions.

Author: Not sure what you are asking here. This is in line with what we have for other grid aggregations so if we change it here we should change for all of them.

            "top_left": "52.4, 4.9",
            "bottom_right": "52.3, 5.0"
          }
        }
      },
      "aggregations": {
        "zoom1": {
          "geohex_grid": {
            "field": "location",
            "precision": 12
          }
        }
      }
    }
  }
}
--------------------------------------------------
// TEST[continued]

[source,console-result]
--------------------------------------------------
{
  ...
  "aggregations": {
    "zoomed-in": {
      "doc_count": 3,
      "zoom1": {
        "buckets": [
          {
            "key": "8c1969c9b2617ff",
            "doc_count": 1
          },
          {
            "key": "8c1969526d753ff",
            "doc_count": 1
          },
          {
            "key": "8c1969526d26dff",
            "doc_count": 1
          }
        ]
      }
    }
  }
}
--------------------------------------------------
// TESTRESPONSE[s/\.\.\./"took": $body.took,"_shards": $body._shards,"hits":$body.hits,"timed_out":false,/]

==== Requests with additional bounding box filtering

The `geohex_grid` aggregation supports an optional `bounds` parameter
that restricts the cells considered to those that intersects the

> bounds provided. The `bounds` parameter accepts the bounding box in

Reviewer: ```suggestion
provided bounds. The `bounds` parameter accepts the bounding box in
```

{
  "aggregations": {
    "tiles-in-bounds": {
      "geohex_grid": {
        "field": "location",
        "precision": 12,
        "bounds": {
          "top_left": "52.4, 4.9",
          "bottom_right": "52.3, 5.0"
        }
      }
    }
  }
}
--------------------------------------------------
// TEST[continued]

[source,console-result]
--------------------------------------------------
{
  ...
  "aggregations": {
    "tiles-in-bounds": {
      "buckets": [
        {
          "key": "8c1969c9b2617ff",
          "doc_count": 1
        },
        {
          "key": "8c1969526d753ff",
          "doc_count": 1
        },
        {
          "key": "8c1969526d26dff",
          "doc_count": 1
        }
      ]
    }
  }
}
--------------------------------------------------
// TESTRESPONSE[s/\.\.\./"took": $body.took,"_shards": $body._shards,"hits":$body.hits,"timed_out":false,/]

==== Options

[horizontal]
field::         Mandatory. The name of the field indexed with GeoPoints.

precision::     Optional. The integer zoom of the key used to define
                cells/buckets in the results. Defaults to 6.
                Values outside of [0,15] will be rejected.

> bounds::         Optional. The bounding box to filter the points in the bucket.

Reviewer: ```suggestion
bounds::
(Optional, object) Bounding box used to filter the geo-points in each bucket.
Accepts the same bounding box formats as the
<<query-dsl-geo-bounding-box-query-accepted-formats,geo-bounding box query>>.
```

> size::          Optional. The maximum number of geohex buckets to return
>                 (defaults to 10,000). When results are trimmed, buckets are
>                 prioritised based on the volumes of documents they contain.

Reviewer: ```suggestion
size::
(Optional, integer) Maximum number of buckets to return. Defaults to 10,000.
When results are trimmed, buckets are prioritized based on the volume of
documents they contain.
```

> shard_size::    Optional. To allow for more accurate counting of the top cells
>                 returned in the final result the aggregation defaults to
>                 returning `max(10,(size x number-of-shards))` buckets from each
>                 shard. If this heuristic is undesirable, the number considered
>                 from each shard can be over-ridden using this parameter.

Reviewer: ```suggestion
shard_size::
(Optional, integer) Number of buckets returned from each shard. Defaults to
`max(10,(size x number-of-shards))` to allow for more a accurate count of the
top cells in the final result.
```
